GH-10083: Apply Nullability to `core` `handler` package

mjd507 opened the pull request:
Related to: #10083

`MessageProcessorMessageSource` now provides a fake message to the underlying messageProcessor since the contract requires.

add `@Nullable` on `CheckedCallable.call()`, since in `LockRequestHandlerAdvice.doInvoke()`, there is a lambda `ExecutionCallback::execute` which will be running inside the `call()`, and its Nullable.

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

mjd507 committed:
GH-10083: Apply Nullability to `core` `handler` package

Related to: #10083

`MessageProcessorMessageSource` now provides a fake message to the underlying messageProcessor since the contract requires.

add `@Nullable` on `CheckedCallable.call()`, since in `LockRequestHandlerAdvice.doInvoke()`, there is a lambda `ExecutionCallback::execute` which will be running inside the `call()`, and its Nullable.

Signed-off-by: Jiandong Ma <jiandong.ma.cn@gmail.com>

artembilan reviewed (changes requested):
That's a lot, but great.
Let's see if we can manage such a big review as well!

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/core/RecoveryCallback.java:
```diff
@@ -35,6 +35,6 @@ public interface RecoveryCallback<T extends @Nullable Object> {
 	 * @return an Object that can be used to replace the callback result that failed
 	 */
-	T recover(AttributeAccessor context, Throwable cause);
+	T recover(AttributeAccessor context, @Nullable Throwable cause);
```
Let's ignore that with `NullAway` for now!
That retry component is going to be reworked soon enough: https://github.com/spring-projects/spring-integration/issues/10345.
After respective Spring Kafka issue is fixed before: https://github.com/spring-projects/spring-kafka/issues/4051.

mjd507 commented on spring-integration-core/src/main/java/org/springframework/integration/support/locks/LockRegistry.java:
```diff
@@ -126,7 +128,7 @@ default <E extends Throwable> void executeLocked(Object lockKey, Duration waitLo
 	 * @since 6.2
 	 */
-	default <T, E extends Throwable> T executeLocked(Object lockKey, Duration waitLockDuration,
+	default <T, E extends Throwable> @Nullable T executeLocked(Object lockKey, Duration waitLockDuration,
```
oh, because I add Nullable on `ExecutionCallback.execute()`,  the implementation is `MethodInvocation.proceed()`, which could return null.

And the caller `LockRequestHandlerAdvice.doInvoke()` call this `lockRegistry.executeLocked(lockKey, callback::execute)`, then build with following error, that's I add `Nullable` on `CheckedCallable.call() `

```java
LockRequestHandlerAdvice.java:165: error: [NullAway] referenced method returns @Nullable, but functional interface method org.springframework.integration.util.CheckedCallable.call() returns @NonNull
                                        return this.lockRegistry.executeLocked(lockKey, waitLockDuration, callback::execute);
                                                                                                          ^
    (see http://t.uber.com/nullaway )
```

mjd507 committed:
revert `@Nullable` on Throwable parameter in `RecoveryCallback.recover()`

revert `@Nullable` on `CheckedCallable.call()` as it is nullable from generic type definition, for `LockRequestHandlerAdvice`, add NullAway since `ExecutionCallback.execute` is Nullable as well.

move fakeMessage to a static final constant in `MessageProcessorMessageSource`

add NullAway on `RequestHandlerRetryAdvice.recoveryCallback.recover()`, need rework on retry component

use `Map.computeIfAbsent` in `PayloadsArgumentResolver` and `PayloadExpressionArgumentResolver`

add more detailed reason (critical path, dataflow analysis) on NullAway method.

Signed-off-by: Jiandong Ma <jiandong.ma.cn@gmail.com>

mjd507 commented on spring-integration-core/src/main/java/org/springframework/integration/core/ErrorMessagePublisher.java:
```diff
@@ -70,7 +70,7 @@ public final void setErrorMessageStrategy(ErrorMessageStrategy errorMessageStrat
 	}
 
-	public final void setChannel(MessageChannel channel) {
+	public final void setChannel(@Nullable MessageChannel channel) {
```
I leave nullable here, one of ctors will pass null to this method.

mjd507 commented on spring-integration-core/src/main/java/org/springframework/integration/handler/ExpressionEvaluatingMessageHandler.java:
```diff
@@ -35,8 +35,7 @@ public class ExpressionEvaluatingMessageHandler extends AbstractMessageHandler {
 
-	@SuppressWarnings("NullAway.Init")
-	private String componentType;
+	private String componentType = "expression-evaluating-handler";
```
yeah, I tried that, unit test failed a lot, there it will call `setComponentType()`

artembilan commented:
Fixed via: https://github.com/spring-projects/spring-integration/commit/a167170aa870d84a79ece1b061d23a8ef17cf1b4.
And subsequent cleanup and improvements: https://github.com/spring-projects/spring-integration/commit/31a7dd496db35de82f1d060754f768ed6554e6c2

Thank you again!
